Firestore/core/test/firebase/firestore/testutil/async_testing.h

```diff
@@ -46,6 +52,84 @@ std::unique_ptr<util::Executor> ExecutorForTesting(const char* name);
  */
 std::shared_ptr<util::AsyncQueue> AsyncQueueForTesting();
 
+constexpr auto kTimeout = std::chrono::seconds(5);
+
+/**
+ * An expected outcome of an asynchronous test.
+ */
+class Expectation {
+ public:
+  Expectation();
+
+  /** Marks this expectation as fulfilled. */
+  void Fulfill();
+
+  /**
+   * Returns a callback function, that when invoked, fullfills the expectation.
+   *
+   * The returned function has a lifetime that's independent of the Expectation
+   * that created it.
+   */
+  std::function<void()> AsCallback() const;
+
+  /**
+   * Returns a `future` that represents the completion of this Expectation.
+   */
+  std::future<void> get_future() const;
+
+ private:
+  std::shared_ptr<std::promise<void>> promise_;
+};
+
+/**
+ * A mixin that supplies utilities for safely writing asynchronous tests.
+ */
+class AsyncTest {
+ public:
+  AsyncTest()
+      : trace_("Test case name",
```

Comment:
Very optional: if you'd like, you can probably move this into an in-class initializer and just have `= default` here.